### PR TITLE
MAINT: next = step

### PR DESCRIPTION
**Pull request by stsievert**
I've realized we need to preserve backwards compatibility #2. This PR is a one line fix:

``` python
next = step = __next__
```

I could rebase to get rid of the commits but I don't think a one-line change is worth it.

### CODE REVIEW

**Comment by andyfaff** [comment]
Thanks for the contribution Scott, do you have any feedback on the general design before I press forward with other minimizers?
I'm going to close this PR because I'm worried about the state of the git repo if I just merge. I'll make the change separately myself.

**Comment by stsievert** [comment]
It looks pretty good from glancing at it. The base class looks to be pretty general.

A couple suggestions:

* In my field we have `n` observations and our model vector has `d` (or maybe `p`) dimensions. This might warrant a change to https://github.com/andyfaff/scipy/commit/a5e50901069bd942746c89b3d1b68d5c138c25d1#diff-c1c2c73ad9e5ddef5673e2e95d19020fR3152
* Could we expose a variable representing the hyperparameters of the optimization? These are variables used in `__next__` and are assumed to be given in papers but up to the practitioner to choose (i.e., `step_size` in gradient descent)
* How could we incorporate a `Function` class to perform minimization over? We could pull the Hessian and gradient from this, and any other relevant parameters. This could help remove some of the keyword args to the `Optimize` class.

> I'm going to close this PR because I'm worried about the state of the git repo if I just merge. I'll make the change separately myself.

Yeah, that's what I would do too

**Comment by andyfaff** [comment]
> In my field we have n observations and our model vector has d (or maybe p) dimensions. This might warrant a change to a5e5090#diff-c1c2c73ad9e5ddef5673e2e95d19020fR3152

That's up for discussion. Remember that the optimizer doesn't know about any observations, they're typically passed in args. Perhaps `dim` is better.

> Could we expose a variable representing the hyperparameters of the optimization? These are variables used in __next__ and are assumed to be given in papers but up to the practitioner to choose (i.e., step_size in gradient descent)

I'm familiar with differential evolution, but am by no means an expert in these optimizers. I'm just intelligently copying and pasting here. We could have a property `hyperparameters` which returns a container with the hyperparameter names. However, I'm not sure what are hyperparameters and which aren't. For example with NelderMead `rho/chi/psi/sigma` are probably hyperparameters (not completely sure what they do without closer reading of the code in __next__), but is there anything else? This feature should be straightforward to add.

> How could we incorporate a `Function` class to perform minimization over? We could pull the Hessian and gradient from this, and any other relevant parameters. This could help remove some of the keyword args to the Optimize class.

Hmmm, have to think a little about this one.

**Comment by stsievert** [comment]
>> How could we incorporate a `Function` class
>
> Hmmm, have to think a little about this one.

I've also spent my free time thinking over the implications, and I think this should be implemented.

1. Implementing this would allow greater flexibility
    * all data necessary to compute could be class parameters
    * it'd be easy for a subclass to register a hook every time `grad` computed (as [torch.autograd.Variable.register_hook])
2. We can preserve backwards compatibility when `jac` or `hess` not specified
3. It would clean up the code
    * the `args` parameter in `minimize.__init__` could go away
    * ...and then `jac` wouldn't need to accept the same args as `fun`

Here's a quick implementation:

``` python
class Function:
    def __init__(self, fn, grad=None, hessian=None):
        self.fn = fn
        self.grad = grad
        self.hessian = hessian

    def __call__(self, *args, **kwargs):
        return self.fn(*args, **kwargs)


class Optimizer:
    def __init__(self, function, grad=None, hess=None, ...):
        # ...
        if grad or hess:
            warn('Change "optimize(fn, ..., grad=grad, hess=hess)" '
                 'to "optimize(Function(fn, grad=grad, hess=hess), ...)"')
```

Similar work:

* PyTorch's function: http://pytorch.org/docs/0.3.0/autograd.html#torch.autograd.Function
* DyNet expression: https://dynet.readthedocs.io/en/latest/python_ref.html#dynet.Expression
* CNTK function: https://cntk.ai/pythondocs/cntk.ops.functions.html#cntk.ops.functions.Function

[torch.autograd.Variable.register_hook]:http://pytorch.org/docs/0.3.0/autograd.html#torch.autograd.Variable.register_hook

**Comment by andyfaff** [comment]
See new commit.

**Comment by andyfaff** [comment]
Probably a good idea to add comments to an issue,
